relay-client/src/client.rs

```diff
@@ -0,0 +1,749 @@
+        let (reply_tx, reply_rx) = oneshot::channel();
+        command_tx
+            .send(Command::GetPendingMessages(reply_tx))
+            .await?;
+        let pending_count = reply_rx.await?;
+        Ok(pending_count)
+    }
+
+    /// Request to replay pending messages
+    pub async fn replay_pending_messages(&self) -> Result<()> {
+        let command_tx = self
+            .command_tx
+            .as_ref()
+            .ok_or_else(|| eyre::eyre!("Client not connected"))?;
+        command_tx.send(Command::ReplayPendingMessages).await?;
+        Ok(())
+    }
+
+    /// Reconnect the client. On restart, pending messages will be replayed.
+    pub async fn reconnect(&self) -> Result<()> {
+        let command_tx = self
+            .command_tx
+            .as_ref()
+            .ok_or_else(|| eyre::eyre!("Client not connected"))?;
+        command_tx.send(Command::Reconnect).await?;
+        Ok(())
+    }
+
+    pub async fn graceful_shutdown(
+        &mut self,
+        wait_for_pending_messages: Duration,
+        wait_for_shutdown: Duration,
+    ) {
+        // Let's wait for all acks to be received
+        if self.has_pending_messages().await.map_or(false, |n| n > 0) {
+            tracing::info!(
+                "Giving {}ms for pending messages to be acked",
+                wait_for_pending_messages.as_millis()
+            );
+            tokio::time::sleep(wait_for_pending_messages).await;
+        }
+        // If there are still pending messages, we retry to send them
+        if self.has_pending_messages().await.map_or(false, |n| n > 0) {
+            tracing::info!("There are still pending messages, replaying...");
+            if let Ok(()) = self.replay_pending_messages().await {
+                tokio::time::sleep(wait_for_pending_messages).await;
+            }
+        }
+
+        // Eventually, there not much more we can do, so we shutdown the client
+        self.shutdown();
+
+        if let Some(shutdown_completed) = self.shutdown_completed.take() {
+            match tokio::time::timeout(wait_for_shutdown, shutdown_completed).await {
+                Ok(_) => tracing::info!("Shutdown completed successfully."),
+                Err(_) => tracing::warn!("Timed out waiting for shutdown to complete."),
+            }
+        }
+    }
+
+    /// Shutdown the client
+    pub fn shutdown(&mut self) {
+        tracing::info!("Shutting down requested");
+        if let Some(token) = self.shutdown_token.take() {
+            token.cancel();
+        }
+    }
+
+    pub async fn wait_for_msg_while_spamming<
+        T: PayloadMatcher,
+        S: IntoPayload + std::clone::Clone,
+    >(
+        &mut self,
+        wait: Duration,
+        spam: S,
+        spam_every: Duration,
+    ) -> Result<T::Output> {
+        let start_time = tokio::time::Instant::now();
+        let mut spam_time = tokio::time::Instant::now();
+        loop {
+            if let Some(payload) = self.check_for_msg::<T>().await {
```

Comment:
suggestion: pretty sure this whole section of code can/should be a loop on a tokio::select

Reply:
Agreed. PR #342 is a refactor of the relay client. I didn't want to break the client as it stands (I didn't write the original)